test/hotspot/jtreg/compiler/lib/template_framework/DataName.java

```diff
@@ -0,0 +1,224 @@
+         */
+        MUTABLE,
+        /**
+         * Used for immutable fields and variables, i.e. writing is not allowed,
+         * for example because the field or variable is final.
+         */
+        IMMUTABLE,
+        /**
+         * When filtering, we sometimes want to indicate that we accept
+         * mutable and immutable fields and variables, for example when
+         * we are only reading, the mutability state does not matter.
+         */
+        MUTABLE_OR_IMMUTABLE
+    }
+
+    /**
+     * Creates a new {@link DataName}.
+     */
+    public DataName {
+    }
+
+    /**
+     * The interface for the type of a {@link DataName}.
+     */
+    public interface Type extends Name.Type {
+        /**
+         * The name of the type, that can be used in code.
+         *
+         * @return The {@link String} representation of the type, that can be used in code.
+         */
+        String name();
+
+        /**
+         * Defines the subtype relationship with other types, which is used to filter {@link DataName}s
+         * in {@link View#exactOf}, {@link View#subtypeOf}, and {@link View#supertypeOf}.
+         *
+         * @param other The other type, where we check if it is the supertype of {@code 'this'}.
+         * @return If {@code 'this'} is a subtype of {@code 'other'}.
+         */
+        boolean isSubtypeOf(DataName.Type other);
+    }
+
+    /**
+     * The {@link View} represents a filtered set of {@link DataName}s in the current scope.
+     * It can be obtained with {@link Template#dataNames}. It can be used to count the
+     * available {@link DataName}s, or sample a random {@link DataName} according to the
+     * weights of the {@link DataName}s in the filtered set.
+     */
+    public static final class View {
+        private final Mutability mutability;
+        private final DataName.Type subtype;
+        private final DataName.Type supertype;
+
+        View(Mutability mutability, DataName.Type subtype, DataName.Type supertype) {
+            this.mutability = mutability;
+            this.subtype = subtype;
+            this.supertype = supertype;
+        }
+
+        View(Mutability mutability) {
+            this(mutability, null, null);
+        }
+
+        NameSet.Predicate predicate() {
+            if (subtype == null && supertype == null) {
+                throw new UnsupportedOperationException("Must first call 'subtypeOf', 'supertypeOf', or 'exactOf'.");
+            }
+            return (Name name) -> {
+                if (!(name instanceof DataName dn)) { return false; }
+                if (mutability == Mutability.MUTABLE && !dn.mutable()) { return false; }
+                if (mutability == Mutability.IMMUTABLE && dn.mutable()) { return false; }
+                if (subtype != null && !dn.type().isSubtypeOf(subtype)) { return false; }
+                if (supertype != null && !supertype.isSubtypeOf(dn.type())) { return false; }
```

Comment:
I suggest to use the full term:
```suggestion
                if (!(name instanceof DataName dataName)) { return false; }
                if (mutability == Mutability.MUTABLE && !dataName.mutable()) { return false; }
                if (mutability == Mutability.IMMUTABLE && dataName.mutable()) { return false; }
                if (subtype != null && !dataName.type().isSubtypeOf(subtype)) { return false; }
                if (supertype != null && !supertype.isSubtypeOf(dataName.type())) { return false; }
```